src/main/kotlin/org/rust/lang/core/ControlFlowGraph.kt

```diff
@@ -0,0 +1,188 @@
+
+typealias CFGGraph = Graph<CFGNodeData, CFGEdgeData>
+typealias CFGNode = Node<CFGNodeData, CFGEdgeData>
+typealias CFGEdge = Edge<CFGNodeData, CFGEdgeData>
+
+class ControlFlowGraph(body: RsBlock) {
+    val graph: CFGGraph = Graph()
+    val owner: RsElement
+    val entry: CFGNode
+    val exit: CFGNode
+
+    private val fnExit: CFGNode
+    private val body: RsBlock
+
+    init {
+        this.entry = graph.addNode(CFGNodeData.Entry)
+        this.owner = body.parent as RsElement
+        this.fnExit = graph.addNode(CFGNodeData.Exit)
+
+        val cfgBuilder = CFGBuilder(graph, entry, fnExit)
+        val bodyExit = cfgBuilder.process(body, entry)
+        cfgBuilder.addContainedEdge(bodyExit, fnExit)
+
+        this.exit = fnExit
+        this.body = body
+    }
+
+    fun isNodeReachable(item: RsElement) = graph.depthFirstTraversal(entry).any { it.data.element == item }
+
+    fun buildLocalIndex(): HashMap<RsElement, MutableList<CFGNode>> {
+        val table = hashMapOf<RsElement, MutableList<CFGNode>>()
+        val func = body.parent
+
+        if (func is RsFunction) {
+            val formals = object : RsVisitor() {
+                override fun visitPat(pat: RsPat) {
+                    table.getOrPut(pat, ::mutableListOf).add(entry)
+                    pat.acceptChildren(this)
+                }
+            }
+
+            func.valueParameters.mapNotNull { it.pat }.forEach { formals.visitPat(it) }
+        }
+
+        graph.forEachNode { node ->
+            val element = node.data.element
+            if (element != null)
+                table.getOrPut(element, ::mutableListOf).add(node)
+        }
+
+        return table
+    }
+
+    fun depthFirstTraversalTrace(): String =
+        graph.depthFirstTraversal(this.entry).map { it.data.text }.joinToString("\n")
+
+
+    /**
+     * Creates graph description written in the DOT language.
+     * Usage: copy the output into `cfg.dot` file and run `dot -Tpng cfg.dot -o cfg.png`
+     */
+    fun createDotDescription(): String {
+        val sb = StringBuilder()
+        sb.append("digraph {\n")
+        graph.forEachEdge { edge ->
+            val source = edge.source
+            val target = edge.target
+            val sourceNode = source.data
+            val targetNode = target.data
+
+            sb.append("    \"${source.index}: ${sourceNode.text}\" -> \"${target.index}: ${targetNode.text}\";\n")
+        }
+        sb.append("}\n")
+        return sb.toString()
+    }
+}
+
+
+sealed class ExitPoint {
+    class Return(val e: RsRetExpr) : ExitPoint()
+    class TryExpr(val e: RsExpr) : ExitPoint() // `?` or `try!`
+    class DivergingExpr(val e: RsExpr) : ExitPoint()
+    class TailExpr(val e: RsExpr) : ExitPoint()
+    class TailStatement(val stmt: RsExprStmt) : ExitPoint()
+
+    companion object {
+        fun process(fn: RsFunction, sink: (ExitPoint) -> Unit) = fn.block?.acceptChildren(ExitPointVisitor(sink))
```

Comment:
I think you should use ordinary body functions instead of this. You don't really need `process` to return `Unit?`